cmd/soroban-rpc/internal/preflight/pool.go

```diff
@@ -48,38 +48,42 @@ func NewPreflightWorkerPool(workerCount uint, ledgerEntryReader db.LedgerEntryRe
 
 func (pwp *PreflightWorkerPool) work() {
 	defer pwp.wg.Done()
-	for {
-		select {
-		case <-pwp.done:
-			return
-		case request := <-pwp.requestChan:
-			preflight, err := GetPreflight(request.ctx, request.params)
-			request.resultChan <- workerResult{preflight, err}
-		}
+	for request := range pwp.requestChan {
+		preflight, err := GetPreflight(request.ctx, request.params)
+		request.resultChan <- workerResult{preflight, err}
 	}
 }
 
 func (pwp *PreflightWorkerPool) Close() {
-	close(pwp.done)
+	if !pwp.isClosed.CompareAndSwap(false, true) {
+		// it was already closed
+		return
+	}
+	close(pwp.requestChan)
 	pwp.wg.Wait()
 }
 
-func (pwp *PreflightWorkerPool) GetPreflight(ctx context.Context, sourceAccount xdr.AccountId, op xdr.InvokeHostFunctionOp) (Preflight, error) {
+var PreflightQueueFullErr = errors.New("preflight queue full")
+
+func (pwp *PreflightWorkerPool) GetPreflight(ctx context.Context, readTx db.LedgerEntryReadTx, sourceAccount xdr.AccountId, op xdr.InvokeHostFunctionOp) (Preflight, error) {
+	if pwp.isClosed.Load() {
```

Comment:
this would not, unfortunately, provide adequate protection against writing to closed channels.
my suggestion is that you would use a bool channel instead, and select on that one.